org.eclipse.lsp4e/src/org/eclipse/lsp4e/LanguageClientImpl.java

```diff
@@ -83,6 +94,51 @@ public final void logMessage(MessageParams message) {
 		CompletableFuture.runAsync(() -> ServerMessageHandler.logMessage(wrapper, message));
 	}
 
+	private Map<String, BlockingQueue<ProgressParams>> progressQueueMap = new HashMap<>();
+
+	@Override
+	public CompletableFuture<Void> createProgress(WorkDoneProgressCreateParams params) {
+		String jobName = params.getToken().getLeft();
+		LinkedBlockingDeque<ProgressParams> blockingQueue = new LinkedBlockingDeque<ProgressParams>();
+		progressQueueMap.put(jobName, blockingQueue);
+
+		Job job = Job.create(jobName, new ICoreRunnable() {
+			@Override
+			public void run(IProgressMonitor monitor) throws CoreException {
+				ProgressParams nextNotification;
+				try {
+					while ((nextNotification = blockingQueue.take()) != null) {
+						WorkDoneProgressNotification nextProgressNotification = nextNotification.getValue().getLeft();
+						if (nextProgressNotification != null) {
+							WorkDoneProgressKind kind = nextProgressNotification.getKind();
+							if (kind == WorkDoneProgressKind.begin) {
+								monitor.beginTask(jobName, 100);
+							} else if (kind == WorkDoneProgressKind.report) {
+								WorkDoneProgressReport report = (WorkDoneProgressReport) nextProgressNotification;
+								monitor.worked(report.getPercentage());
+							} else if (kind == WorkDoneProgressKind.end) {
+								monitor.worked(100);
+								return;
+							}
+						}
+					}
+				} catch (InterruptedException e) {
+					LanguageServerPlugin.logError(e);
+					Thread.currentThread().interrupt();
+				}
+			}
+		});
+		job.schedule();
+		return CompletableFuture.completedFuture(null);
```

Comment:
Cancellations would be a nice addition :)

Reply:
done